File: rust/496.next-greater-element-i.rs
/*
 * @lc app=leetcode id=496 lang=rust
 *
 * [496] Next Greater Element I
 */

// @lc code=start
impl Solution {
    pub fn next_greater_element(nums1: Vec<i32>, nums2: Vec<i32>) -> Vec<i32> {
        use std::collections::HashMap;
        let mut greater = HashMap::from([(nums2.last().unwrap(), -1)]);
        for i in (0..nums2.len() - 1).rev() {
            let mut n = nums2[i + 1];
            if nums2[i] > n {
                while n != -1 && n < nums2[i] {
                    n = greater[&n];
                }
            }
            greater.insert(&nums2[i], n);
        }
        nums1.iter().map(|n| greater[n]).collect()
    }
}
// @lc code=end

File: javascript/496.next-greater-element-i.js
/*
 * @lc app=leetcode id=496 lang=javascript
 *
 * [496] Next Greater Element I
 */

// @lc code=start
/**
 * @param {number[]} nums1
 * @param {number[]} nums2
 * @return {number[]}
 */
var nextGreaterElement = function (nums1, nums2) {
  const greater = {
    [nums2[nums2.length - 1]]: -1,
  };
  for (let i = nums2.length - 2; i >= 0; i--) {
    let n = nums2[i + 1];
    if (nums2[i] > n) {
      while (n !== -1 && n < nums2[i]) n = greater[n];
    }
    greater[nums2[i]] = n;
  }
  return nums1.map((n) => greater[n]);
};
// @lc code=end
